lib/jekyll/document.rb

```diff
@@ -273,6 +273,11 @@ def read(opts = {})
       end
     end
 
+    # The Renderer instance for this document
+    def renderer(payload = nil)
+      @renderer ||= Jekyll::Renderer.new(site, self, payload)
```

Comment:
By accepting the payload but memoizing, you're breaking a contract with the user that the input will affect the return value.

```text
>> doc.renderer
#<Jekyll::Renderer@12345 @payload=nil>
>> doc.renderer(my_payload)
#<Jekyll::Renderer@12345 @payload=nil>
>> !!! I should get back a renderer that has `my_payload` set, not nil!
```

One possible solution to this is to:

```ruby
def renderer(payload = nil)
  @renderer ||= Jekyll::Renderer.new(site, self, payload)
  @renderer.tap { |r| r.payload = payload }
end
```

This will return the Renderer instance with the given payload.

Reply:
nice..  :+1: